### ISSUE TITLE
Shortcode Scratch variable not rendering on local server after edit to the Parent Page template

### ISSUE THREAD

**Issue by onedrawingperday** [gohugoio/hugo#8255]
This [report](https://discourse.gohugo.io/t/page-scratch-missing-on-rebuilds-in-live-reload/31206 ) came up in the forum and I can confirm that a Scratch variable set in a Shortcode does not render when called from the template of its Parent Page.

The Scratch variable is rendered on first load of the local server in fast rendering mode.

However if `--disabelFastRender` is used then the variable is not rendered.

Also the variable is not rendered when the command `hugo` is executed.

Repo to reproduce: https://github.com/Oliver-Fish/hugo-scratchpad

---

Since Shortcodes are rendered first, I would expect a Scratchpad set in a Shortcode to be available from a Parent Page's template.

cc: @Oliver-Fish

**Comment by bep** [comment]
>Since Shortcodes are rendered first, I would expect a Scratchpad set in a Shortcode to be available from a Parent Page's template.

Sure, but since the shortcodes gets lazily rendered, you need to invoke `.Content` (or one of the other ´Content` methods, e.g. `WordCount`) first.

**Comment by onedrawingperday** [comment]
> you need to invoke .Content (or one of the other ´Contentmethods, e.g.WordCount`) first.

Thank you for the reply.

With the above invocation the Scratch variable does render when the project is published with the command `hugo`.

Also the Scratch variable now renders on first load of both `hugo server` and `hugo server --disableFastRender`

However once changes are made then the Scratch variable is again missing once LiveReload refreshes the browser.

Is this due to the way the rendering order changes with LiveReload? Or can it be fixed?  

Sorry to bother you with these questions.

The person who reported this issue is a Go developer and  has offered himself to look into this  issue if it can be fixed.

Although it now seems like a limitation to me. 

---

By the way the use case for the above is the rendering of [a FAQ schema](https://developers.google.com/search/docs/data-types/faqpage) through Nested Shortcodes, simliar to how [Gutenberg blocks](https://ultimategutenberg.com/faq/?utm_source=wp-repo&utm_medium=link&utm_campaign=readme) function in Wordpress)

**Comment by bep** [comment]
>With the above invocation the Scratch variable does render when the project is published with the command hugo.

Well, I'm just looking at the template code and I'm saying that

1. The value in scratch does not exist until its somehow added
2. According to the template code it is accessed before it is written

I suggest adjusting fixing the template first.

**Comment by onedrawingperday** [comment]
Before posting my relpy above I adjusted the template to:
```
<body>
    {{ $content := .Content }}
    <h1>{{ .Scratch.Get "question" }}</h1>
   {{ $content }}
</body>
```
Simply calling `.Content` though a variable does make the Scratch available, without changing the order of the markup elements.

However the LiveReload issue on the local server persist, as noted above.

**Comment by Oliver-Fish** [comment]
I have updated the proof of concept repo to correct the issue with the template, this has actually changed the steps to re-create the issue slightly.

The scratch variable is now only empty if I trigger a live reload by editing layouts/test/single.html

So to be clear I did three tests

Edited layouts/test/single.html - Empty
Edited layouts/shortcodes/test.html - Value as expected
Edited content/test.md - Value as expected

Additionally, if I edit single.html I will get the issue but if I then edit test.html or test.md triggering another reload the issue is corrected.

So any change to single.html causes the issue but it can be reset by editing any of the other monitored files. 

I am not familiar with the Hugo codebase however, a naive guess would be that the layout modification appears to not be following the same re-rendering steps as the other files.

**Comment by onedrawingperday** [comment]
>Edited layouts/test/single.html - Empty
Edited layouts/shortcodes/test.html - Value as expected
Edited content/test.md - Value as expected

I changed the title of the issue to make it clearer of what the issue is about.

Basically if the Shortcode parent page template is edited and LiveReload refreshes the page, then the Scratch variable is not rendering.

If this can be fixed it would be nice to have.

But since it does not interrupt the content editing workflow, I can live with it. 
(It's not often one has to make changes in the template after a project has been developed).

Anyway, again thanks for the guidance. 

I was not aware that Shortcodes are lazy-loaded and that the `Content` variable has to be called beforehand in the template.

**Comment by Oliver-Fish** [comment]
> If this can be fixed it would be nice to have.

Agreed this unblocks me, my site is new so I am still going to hit this pain often but it won't last forever.
  
> I was not aware that Shortcodes are lazy-loaded and that the `Content` variable has to be called beforehand in the template.

This wasn't clear from the docs, maybe we can add this somewhere? it could be quite a timesink of a bug to get stuck on this problem without knowing this.

@bep

Is there a reason why Shortcodes are lazy-loaded like this? Isn't this the more common use-case that we should optimize for as this can be quite confusing?

But of course, don't know the reason this was done so I might be missing something here.

Either way, thank you both @onedrawingperday and @bep really great work on managing issues and support for Hugo this is above the standard I have seen most large enterprise support contracts offer.

**Comment by onedrawingperday** [comment]
> Is there a reason why Shortcodes are lazy-loaded like this? Isn't this the more common use-case that we should optimize for as this can be quite confusing?

I'm not the maintainer, but I would guess that the lazy loading is use for performance reasons.

**Comment by Oliver-Fish** [comment]
> > Is there a reason why Shortcodes are lazy-loaded like this? Isn't this the more common use-case that we should optimize for as this can be quite confusing?
> 
> I'm not the maintainer, but I would guess that the lazy loading is used for performance reasons.

But isn't the common use-case to always have a content call in your file at some point? So doing it up-front, regardless of where it's called in the file, improves the UX for the developer as it prevents nasty bugs like this one. 

However, I could be missing a use-case where this is valid but it'd be good to understand what that is.

**Comment by onedrawingperday** [comment]
>But isn't the common use-case to always have a content call in your file at some point?

Not necessarily. I do maintain a couple of e-commerce projects that rely on front matter parameters and do not call the `.Content` variable in the templates.

**Comment by Oliver-Fish** [comment]
> > But isn't the common use-case to always have a content call in your file at some point?
> 
> Not necessarily. I do maintain a couple of e-commerce projects that rely on front matter parameters and do not call the `.Content` variable in the templates.

Intresting, and in these cases if you called Content would it actually contain any data?

**Comment by onedrawingperday** [comment]
>Interesting, and in these cases if you called .Content would it actually contain any data?

No, `.Content` returns empty (not even <nil> gets printed). Anyway the forum is meant for discussion, not this issue.
